lib/Data/Time/Clock/POSIX.hs

```diff
@@ -1,27 +1,117 @@
 -- | POSIX time, if you need to deal with timestamps and the like.
 -- Most people won't need this module.
 module Data.Time.Clock.POSIX
-(
-    posixDayLength,POSIXTime,posixSecondsToUTCTime,utcTimeToPOSIXSeconds,getPOSIXTime,getCurrentTime
-) where
+    ( -- * Raw POSIX timestamps
+      POSIXTimeRaw(..)
+    , normalizeRaw
+    , posixRawToPOSIX
+    , posixRawToUTCTime
+    , getPOSIXTimeRaw
+    , posixDayLengthRaw
+     -- * high precision POSIX time type
+    , posixDayLength
+    , POSIXTime
+    , unixEpochDay
+    , posixSecondsToUTCTime
+    , utcTimeToPOSIXSeconds
+    , getPOSIXTime
+    , getCurrentTime
+    ) where
 
 import Data.Time.Clock.UTC
+import Data.Time.Clock.Scale (picosecondsToDiffTime)
 import Data.Time.Calendar.Days
 import Data.Fixed
 import Control.Monad
+import Data.Int (Int64)
 
 #include "HsTimeConfig.h"
 
 #ifdef mingw32_HOST_OS
-import Data.Word    ( Word64)
+import Data.Word (Word64)
 import System.Win32.Time
 #elif HAVE_CLOCK_GETTIME
 import Data.Time.Clock.CTimespec
-import Foreign.C.Types (CTime(..))
+import Foreign.C.Types (CTime(..), CLong(..))
 #else
 import Data.Time.Clock.CTimeval
+import Foreign.C.Types (CLong(..))
 #endif
 
+--------------------------------------------------------------------------------
+
+data POSIXTimeRaw = POSIXTimeRaw
+    { ptSeconds ::     {-# UNPACK #-} !Int64
+    , ptNanoSeconds :: {-# UNPACK #-} !Int64
```

Comment:
is `ptNanoSeconds` ever negative? Would `Word64` be a better type? Or even `Word32`?